torch/_inductor/codegen/cpp.py

```diff
@@ -2158,37 +2158,46 @@ def is_bf16_scheduler(self, scheduler_node: SchedulerNode):
 
     def legalize_bf16(self, nodes):
         def add_to_dtype(sub_graph: torch.fx.Graph):
-            for node in sub_graph.nodes:
-                _node: torch.fx.Node = node
-                if _node.target in ["load", "constant"]:
-                    assert len(_node.args) == 3
-                    ops = _node.args[0]
-                    # If the node is constant, the last arg is dtype
-                    load_dtype = (
-                        V.graph.get_dtype(_node.args[1])
-                        if _node.target == "load"
-                        else _node.args[-1]
-                    )
+            def is_bf16_load_or_constant(node: torch.fx.Node):
+                if node.target not in ["load", "constant"]:
+                    return False
+                assert len(node.args) == 3
+                # If the node is constant, the last arg is dtype
+                load_dtype = (
+                    V.graph.get_dtype(node.args[1])
+                    if node.target == "load"
+                    else node.args[-1]
+                )
+                return load_dtype == torch.bfloat16
 
-                    if load_dtype == torch.bfloat16:
-                        with sub_graph.inserting_after(_node):
-                            to_type_node = sub_graph.call_method(
-                                "to_dtype", args=(ops, _node, torch.float)
-                            )
-                            to_type_node_args = to_type_node.args
-                            _node.replace_all_uses_with(to_type_node)
-                            to_type_node.args = to_type_node_args
-                            metrics.cpp_to_dtype_count += 1
-                elif _node.target == "store":
-                    ops, store_var, _, value_var, _ = _node.args
-                    store_dtype = V.graph.get_dtype(store_var)
-                    if store_dtype == torch.bfloat16:
-                        with sub_graph.inserting_before(_node):
-                            to_type_node = sub_graph.call_method(
-                                "to_dtype", args=(ops, value_var, torch.bfloat16)
-                            )
-                            _node.replace_input_with(value_var, to_type_node)
-                            metrics.cpp_to_dtype_count += 1
+            def is_bf16_store(node: torch.fx.Node):
+                if node.target != "store":
+                    return False
+                _, store_var, _, _, _ = node.args
+                store_dtype = V.graph.get_dtype(store_var)
+                return store_dtype == torch.bfloat16
+
+            sub_graph_nodes = sub_graph.nodes
+            to_bf16_legalized_nodes = []
+            for _node in sub_graph_nodes:
+                if is_bf16_load_or_constant(_node):
+                    ops = _node.args[0]
+                    with sub_graph.inserting_after(_node):
```

Comment:
Is it safe to modify the node list `subgraph.nodes` inplace? Should you also record the node list to legalize with and do the graph rewrite after the loop?

Reply:
Changed to a list to avoid in place change the graph nodes.